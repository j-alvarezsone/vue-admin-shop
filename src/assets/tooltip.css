:root {
  --tooltip-display: none;
  --tooltip-text-align: start;
  --tooltip-fontSize: 1rem;
  --tooltip-lineHeight: 1rem;
  --tooltip-backgroundColor: #1e293b;
  --tooltip-textColor: #ffffff;
  --tooltip-padding: 0.5rem 0.75rem;
  --tooltip-borderRadius: 0.5rem;
  --tooltip-position: absolute;
  --tooltip-zIndex: 9999;
  --tooltip-width: max-content;
  --tooltip-maxWidth: 256px;
  --tooltip-maxHeight: 800px;
  --tooltip-boxShadow: 0px 12px 16px -4px rgba(16, 24, 40, 0.08), 0px 4px 6px -2px rgba(16, 24, 40, 0.03);
  /* Arrow */
  --tooltip-arrowBorderColor: #1e293b transparent transparent transparent;
  --tooltip-arrow-display: none;
}

.tooltip {
  display: var(--tooltip-display);
  text-align: var(--tooltip-text-align);
  font-size: var(--tooltip-fontSize);
  line-height: var(--tooltip-lineHeight);
  background-color: var(--tooltip-backgroundColor);
  color: var(--tooltip-textColor);
  padding: var(--tooltip-padding);
  border-radius: var(--tooltip-borderRadius);
  position: var(--tooltip-position);
  z-index: var(--tooltip-zIndex);
  width: var(--tooltip-width);
  max-width: var(--tooltip-maxWidth);
  max-height: var(--tooltip-maxHeight);
  box-shadow: var(--tooltip-boxShadow);
  overflow-wrap: break-word;
  line-height: 1.5rem;
  overflow-y: auto;
}

.tooltip-arrow {
  display: var(--tooltip-arrow-display);
  position: absolute;
  width: 0;
  height: 0;
  border-style: solid;
  border-width: 6px;
  border-color: var(--tooltip-arrowBorderColor);
  z-index: var(--tooltip-zIndex);
}
