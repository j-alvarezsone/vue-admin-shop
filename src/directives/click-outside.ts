import type { DirectiveBinding } from "vue";

type ClickableElement = HTMLElement & {
  clickOutsideEvent?: (event: Event) => void
};

export const clickOutside = {
  mounted(el: ClickableElement, binding: DirectiveBinding) {
    if (typeof binding.value === "function") {
      el.clickOutsideEvent = (event) => {
        const targetElement = event.target as Element;
        if (!(el === targetElement || el.contains(targetElement))) {
          binding.value(event);
        }
      };
    } else {
      el.clickOutsideEvent = (event: Event) => {
        const targetElement = event.target as Element;
        const { handler, ignore } = binding.value;
        const ignoreElement = ignore?.some((i: any) => {
          return (
            targetElement.classList.contains(i) || targetElement.id === i || targetElement === i
          );
        });

        if (!ignoreElement && !(el === targetElement || el.contains(targetElement))) {
          handler(event);
        }
      };
    }
    document.addEventListener("click", el.clickOutsideEvent);
  },
  unmounted(el: ClickableElement) {
    if (el.clickOutsideEvent) {
      document.body.removeEventListener("click", el.clickOutsideEvent);
    }
  },
};
