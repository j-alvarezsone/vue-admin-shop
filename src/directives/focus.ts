import type { Directive, DirectiveBinding } from "vue";

export const focus: Directive = {
  mounted: (el: HTMLElement, binding: DirectiveBinding): void => {
    const { value } = binding;

    if (value) {
      el.focus();
    }
  },
};
