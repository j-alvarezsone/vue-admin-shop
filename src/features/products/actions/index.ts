export * from "./create-update-product";
export * from "./get-product-by-id";
export * from "./get-product-image";
export * from "./get-products";
