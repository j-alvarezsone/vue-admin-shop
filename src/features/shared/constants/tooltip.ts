export const TOOLTIP = {
  DISPLAY: "--tooltip-display",
  TOP: "--tooltip-top",
  BOTTOM: "--tooltip-bottom",
  LEFT: "--tooltip-left",
  RIGHT: "--tooltip-right",
  MAX_WIDTH: "--tooltip-maxWidth",
  MAX_HEIGHT: "--tooltip-maxHeight",
  FONT_SIZE: "--tooltip-fontSize",
  BACKGROUND_COLOR: "--tooltip-backgroundColor",
  TEXT_COLOR: "--tooltip-textColor",
  PADDING: "--tooltip-padding",
  Z_INDEX: "--tooltip-zIndex",
  ARROW_BORDER_COLOR: "--tooltip-arrowBorderColor",
  ARROW_DISPLAY: "--tooltip-arrow-display",
} as const;

export const TOOLTIP_PLACEMENTS = {
  TOP: "top",
  TOP_START: "topStart",
  TOP_END: "topEnd",
  BOTTOM: "bottom",
  BOTTOM_START: "bottomStart",
  BOTTOM_END: "bottomEnd",
  LEFT: "left",
  LEFT_START: "leftStart",
  LEFT_END: "leftEnd",
  RIGHT: "right",
  RIGHT_START: "rightStart",
  RIGHT_END: "rightEnd",
} as const;
