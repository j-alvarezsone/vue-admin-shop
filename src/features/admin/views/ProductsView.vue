<script setup lang="ts">
import type { Size } from "@/features/products/types/product";
import { getProductsAction } from "@/features/products/actions";
import { deleteProductAction } from "@/features/products/actions/delete-product";
import Badge from "@/features/shared/components/ui/Badge.vue";
import Button from "@/features/shared/components/ui/Button.vue";
import InputSelect from "@/features/shared/components/ui/form/InputSelect.vue";
import InputText from "@/features/shared/components/ui/form/InputText.vue";
import Pagination from "@/features/shared/components/ui/Pagination.vue";
import usePagination from "@/features/shared/composables/usePagination";
import { useMutation, useQuery, useQueryClient } from "@tanstack/vue-query";
import { computed, ref, watchEffect } from "vue";
import { useToast } from "vue-toastification";
import AdminProductsSkeleton from "../components/AdminProductsSkeleton.vue";

const queryClient = useQueryClient();
const { page } = usePagination();
const toast = useToast();

const search = ref<string>("");
const size = ref<Size | "all">("all");
const sizeOptions = [
  { value: "all", label: "All" },
  { value: "XS", label: "XS" },
  { value: "S", label: "S" },
  { value: "M", label: "M" },
  { value: "L", label: "L" },
  { value: "XL", label: "XL" },
  { value: "XXL", label: "XXL" },
];

const { data: products, isLoading, isFetching } = useQuery({
  queryKey: ["products", { page }],
  queryFn: () => getProductsAction(page.value),
  staleTime: 1000 * 30,
});

const { mutate } = useMutation({
  mutationFn: deleteProductAction,
  onSuccess: () => {
    queryClient.invalidateQueries({ queryKey: ["products", { page }] });
    toast.success("Product deleted successfully");
  },
  onError: () => {
    toast.error("Failed to delete product");
  },
});

const filteredProducts = computed(() => {
  if (!products.value || !size.value) { return products.value; }

  if (size.value !== "all") {
    return products.value.filter((product) => product.sizes.includes(size.value as Size));
  }

  return products.value.filter((product) =>
    product.title.toLowerCase().includes(search.value.toLowerCase()),
  );
});

watchEffect(() => {
  queryClient.prefetchQuery({
    queryKey: ["products", { page: page.value + 1 }],
    queryFn: () => getProductsAction(page.value + 1),
    staleTime: 1000 * 30,
  });
});
</script>

<template>
  <div class="bg-white px-5 py-2 rounded">
    <div class="flex items-center gap-6">
      <h1 class="text-3xl">
        Products
      </h1>
      <InputText v-model="search" class="w-full" name="search" icon="search" placeholder="Search by title" clearable />
      <InputSelect v-model="size" :options="sizeOptions" class="w-60" name="size" placeholder="Select size" clearable />
    </div>
    <div class="py-8 w-full">
      <div class="shadow overflow-hidden rounded border-b border-gray-200">
        <table class="min-w-full bg-white">
          <thead class="bg-gray-800 text-white">
            <tr>
              <th class="w-10 text-left py-3 px-4 uppercase font-semibold text-sm">
                Images
              </th>
              <th class="flex-1 text-left py-3 px-4 uppercase font-semibold text-sm">
                Title
              </th>
              <th class="w-28 py-3 px-4 uppercase font-semibold text-sm">
                Price
              </th>
              <th class="w-60 text-left py-3 px-4 uppercase font-semibold text-sm">
                Sizes
              </th>
            </tr>
          </thead>
          <AdminProductsSkeleton v-if="isLoading || isFetching" />
          <tbody v-else class="text-gray-700">
            <tr
              v-for="product in filteredProducts"
              :key="product.id"
              class="even:bg-gray-100 odd:bg-white hover:bg-gray-200"
            >
              <td class="text-left py-3 px-4">
                <img :src="product.images[0]" :alt="product.title" class="size-10 object-cover">
              </td>
              <td class="text-left py-3 px-4">
                <RouterLink class="hover:text-blue-500 hover:underline underline-offset-4" :to="{ name: 'admin-product', params: { productId: product.id } }">
                  {{ product.title }}
                </RouterLink>
              </td>
              <td class="text-center py-3 px-4">
                <span class="font-semibold centered">
                  <Badge size="sm">
                    ${{ product.price }}
                  </Badge>
                </span>
              </td>
              <td class="text-left py-3 px-4 flex  items-center justify-between">
                <span class="text-sm text-gray-500">
                  {{ product.sizes.join(", ") }}
                </span>
                <Button v-tooltip="'Delete'" icon="trash" variant="transparent" icon-size="size-6" @click="mutate(product.id)" />
              </td>
            </tr>
          </tbody>
        </table>
        <div v-if="!filteredProducts?.length" class="h-80 w-full centered">
          <span class="text-2xl">No products found...</span>
        </div>
      </div>
    </div>
    <Pagination
      class="bg-white"
      :has-more-data="!!products && products.length < 10"
      :page
    />
  </div>
</template>
