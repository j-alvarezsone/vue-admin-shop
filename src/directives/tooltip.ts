import type { Directive, DirectiveBinding } from "vue";
import { TOOLTIP, TOOLTIP_PLACEMENTS } from "@/features/shared/constants/tooltip";

const tooltipContainer = document.createElement("div");
let hideTimeout: number | null = null;
let wheelEventHandler: (() => void) | null = null;
let arrow: HTMLElement | null = null;

export const tooltip: Directive = {
  beforeMount(el: HTMLElement, binding: DirectiveBinding) {
    const mouseEnterHandler = () => handleMouseEnter(el, tooltipContainer, binding);
    const mouseLeaveHandler = () => handleMouseLeave(el, tooltipContainer, binding);

    el.onmouseenter = mouseEnterHandler;
    el.onmouseleave = mouseLeaveHandler;
  },
  updated(el: HTMLElement, binding: DirectiveBinding) {
    const mouseEnterHandler = () => handleMouseEnter(el, tooltipContainer, binding);
    const mouseLeaveHandler = () => handleMouseLeave(el, tooltipContainer, binding);

    el.onmouseenter = mouseEnterHandler;
    el.onmouseleave = mouseLeaveHandler;
  },
  beforeUnmount(el: HTMLElement) {
    el.onmouseenter = null;
    el.onmouseleave = null;
    tooltipContainer.onmouseenter = null;
    tooltipContainer.onmouseleave = null;

    if (document.body.contains(tooltipContainer)) {
      document.body.removeChild(tooltipContainer);
    }

    removeArrow();
  },
};

function updateTooltip(el: HTMLElement, { value, modifiers }: DirectiveBinding, container: HTMLDivElement) {
  if (typeof value === "string" || typeof value === "number") {
    const stringValue = typeof value === "number" ? value.toString() : value;

    if (modifiers.html) {
      container.innerHTML = stringValue;
      container.style.setProperty(TOOLTIP.PADDING, "0.75rem");
    } else {
      container.textContent = stringValue;
    }
  } else if (typeof value === "object") {
    applyObjectTooltipStyles(value, container);
  }

  container.className = "";
  container.classList.add("tooltip");

  setPlacement(el, container, modifiers, value);

  if (modifiers.bgLight || value?.bgLight) {
    container.style.setProperty(TOOLTIP.BACKGROUND_COLOR, "white");
    container.style.setProperty(TOOLTIP.TEXT_COLOR, "#374151");
  }

  if (modifiers.noArrow || value?.noArrow) {
    removeArrow();
  }
}

function applyObjectTooltipStyles(value: any, container: HTMLDivElement) {
  if (value?.html) {
    container.innerHTML = value.content;
    container.style.setProperty(TOOLTIP.PADDING, "0.75rem");
  } else {
    container.textContent = value.content;
  }

  if (value?.maxWidth) {
    container.style.setProperty(TOOLTIP.MAX_WIDTH, value.maxWidth);
  }

  if (value?.maxHeight) {
    container.style.setProperty(TOOLTIP.MAX_HEIGHT, value.maxHeight);
  }

  if (value?.fontSize) {
    container.style.setProperty(TOOLTIP.FONT_SIZE, value.fontSize);
  }

  if (value?.padding) {
    container.style.setProperty(TOOLTIP.PADDING, value.padding);
  }

  if (value?.zIndex) {
    container.style.setProperty(TOOLTIP.Z_INDEX, value.zIndex);
    arrow?.style.setProperty(TOOLTIP.Z_INDEX, value.zIndex);
  }
}

function setPlacement(
  el: HTMLElement,
  container: HTMLDivElement,
  modifiers: DirectiveBinding["modifiers"],
  value: DirectiveBinding["value"],
) {
  switch (value?.placement || true) {
    case modifiers.left:
    case TOOLTIP_PLACEMENTS.LEFT:
      alignLeft(el, container, modifiers, value);
      isOutOfBounds(container, (isOut) => {
        if (isOut) {
          alignRight(el, container, modifiers, value);
        }
      });
      break;
    case modifiers.leftStart:
    case TOOLTIP_PLACEMENTS.LEFT_START:
      alignLeftStart(el, container, modifiers, value);
      isOutOfBounds(container, (isOut) => {
        if (isOut) {
          alignRightStart(el, container, modifiers, value);
        }
      });
      break;
    case modifiers.leftEnd:
    case TOOLTIP_PLACEMENTS.LEFT_END:
      alignLeftEnd(el, container, modifiers, value);
      isOutOfBounds(container, (isOut) => {
        if (isOut) {
          alignRightEnd(el, container, modifiers, value);
        }
      });
      break;
    case modifiers.right:
    case TOOLTIP_PLACEMENTS.RIGHT:
      alignRight(el, container, modifiers, value);
      isOutOfBounds(container, (isOut) => {
        if (isOut) {
          alignLeft(el, container, modifiers, value);
        }
      });
      break;
    case modifiers.rightStart:
    case TOOLTIP_PLACEMENTS.RIGHT_START:
      alignRightStart(el, container, modifiers, value);
      isOutOfBounds(container, (isOut) => {
        if (isOut) {
          alignLeftStart(el, container, modifiers, value);
        }
      });
      break;
    case modifiers.rightEnd:
    case TOOLTIP_PLACEMENTS.RIGHT_END:
      alignRightEnd(el, container, modifiers, value);
      isOutOfBounds(container, (isOut) => {
        if (isOut) {
          alignLeftEnd(el, container, modifiers, value);
        }
      });
      break;
    case modifiers.bottom:
    case TOOLTIP_PLACEMENTS.BOTTOM:
      alignBottom(el, container, modifiers, value);
      isOutOfBounds(container, (isOut) => {
        if (isOut) {
          alignTop(el, container, modifiers, value);
        }
      });
      break;
    case modifiers.bottomStart:
    case TOOLTIP_PLACEMENTS.BOTTOM_START:
      alignBottomStart(el, container, modifiers, value);
      isOutOfBounds(container, (isOut) => {
        if (isOut) {
          alignTopStart(el, container, modifiers, value);
        }
      });
      break;
    case modifiers.bottomEnd:
    case TOOLTIP_PLACEMENTS.BOTTOM_END:
      alignBottomEnd(el, container, modifiers, value);
      isOutOfBounds(container, (isOut) => {
        if (isOut) {
          alignTopEnd(el, container, modifiers, value);
        }
      });
      break;
    case modifiers.topStart:
    case TOOLTIP_PLACEMENTS.TOP_START:
      alignTopStart(el, container, modifiers, value);
      isOutOfBounds(container, (isOut) => {
        if (isOut) {
          alignBottomStart(el, container, modifiers, value);
        }
      });
      break;
    case modifiers.topEnd:
    case TOOLTIP_PLACEMENTS.TOP_END:
      alignTopEnd(el, container, modifiers, value);
      isOutOfBounds(container, (isOut) => {
        if (isOut) {
          alignBottomEnd(el, container, modifiers, value);
        }
      });
      break;
    case TOOLTIP_PLACEMENTS.TOP:
      alignTop(el, container, modifiers, value);
      isOutOfBounds(container, (isOut) => {
        if (isOut) {
          alignBottom(el, container, modifiers, value);
        }
      });
      break;
    default:
      alignTop(el, container, modifiers, value);
      isOutOfBounds(container, (isOut) => {
        if (isOut) {
          alignBottom(el, container, modifiers, value);
        }
      });
  }
}

function removeTooltip(container: HTMLDivElement, binding: DirectiveBinding) {
  if (typeof binding.value === "string" || typeof binding.value === "object" || typeof binding.value === "number") {
    clearTooltip(binding, container);
  }
}

function removeArrow() {
  if (arrow && document.body.contains(arrow)) {
    document.body.removeChild(arrow);
    arrow.className = "";
    arrow.style.cssText = "";
    arrow = null;
  }
}

function handleMouseEnter(el: HTMLElement, container: HTMLDivElement, binding: DirectiveBinding) {
  const isValueMissing = !binding.value;
  const isContentMissing = typeof binding.value === "object" && !binding.value.content;

  if (isValueMissing || isContentMissing || binding.value?.disabled) {
    clearTooltip(binding, container);
    return;
  }

  if (hideTimeout) {
    clearTimeout(hideTimeout);
    hideTimeout = null;
  }

  if (arrow) {
    removeArrow();
  }

  container.style.setProperty(TOOLTIP.DISPLAY, "block");
  document.body.appendChild(container);

  arrow = document.createElement("div");
  arrow.className = "tooltip-arrow";
  arrow?.style.setProperty(TOOLTIP.ARROW_DISPLAY, "block");
  document.body.appendChild(arrow);

  updateTooltip(el, binding, container);

  wheelEventHandler = () => handleMouseLeave(el, container, binding);
  window.addEventListener("wheel", wheelEventHandler);

  container.onmouseenter = () => {
    if (hideTimeout) {
      clearTimeout(hideTimeout);
      hideTimeout = null;
    }
  };
}

function handleMouseLeave(el: HTMLElement, container: HTMLDivElement, binding: DirectiveBinding) {
  if (binding.modifiers.autoHide || binding.value?.autoHide) {
    removeTooltip(container, binding);
  } else {
    hideTimeout = window.setTimeout(() => {
      removeTooltip(container, binding);
    }, 50);
  }

  if (wheelEventHandler) {
    window.removeEventListener("wheel", wheelEventHandler);
    wheelEventHandler = null;
  }

  container.onmouseleave = () => handleMouseLeave(el, container, binding);
}

function clearTooltip({ modifiers, value, arg }: DirectiveBinding, container: HTMLDivElement) {
  container.className = "";
  container.style.cssText = "";

  if (arrow) {
    removeArrow();
  }

  if (modifiers.html || value?.html) {
    container.innerHTML = "";
  } else {
    container.textContent = "";
  }

  if (arg) {
    arg = undefined;
  }
}

function getElRect(el: HTMLElement) {
  return el.getBoundingClientRect();
}

function setPosition(container: HTMLElement, left: number, top: number) {
  container.style.left = `${left}px`;
  container.style.top = `${top}px`;
}

function setArrowTopPosition(
  el: HTMLElement,
  arrow: HTMLElement,
  modifiers: DirectiveBinding["modifiers"],
  value: DirectiveBinding["value"],
) {
  const elRect = getElRect(el);

  arrow.style.setProperty(TOOLTIP.ARROW_BORDER_COLOR, "#1e293b transparent transparent transparent");

  if (modifiers.bgLight || value?.bgLight) {
    arrow.style.setProperty(TOOLTIP.ARROW_BORDER_COLOR, "white transparent transparent transparent");
  }

  let arrowLeft = elRect.left + elRect.width / 2 - arrow.offsetWidth / 2;
  const arrowTop = elRect.top - arrow.offsetHeight;

  if (modifiers.topStart || value?.placement === TOOLTIP_PLACEMENTS.TOP_START) {
    arrowLeft = elRect.left + arrow.offsetWidth;
  }

  if (modifiers.topEnd || value?.placement === TOOLTIP_PLACEMENTS.TOP_END) {
    arrowLeft = elRect.right - arrow.offsetWidth * 2;
  }

  setPosition(arrow, arrowLeft, arrowTop);
}

function setArrowBottomPosition(
  el: HTMLElement,
  arrow: HTMLElement,
  modifiers: DirectiveBinding["modifiers"],
  value: DirectiveBinding["value"],
) {
  const elRect = getElRect(el);

  arrow.style.setProperty(TOOLTIP.ARROW_BORDER_COLOR, "transparent transparent #1e293b transparent");

  if (modifiers.bgLight || value?.bgLight) {
    arrow.style.setProperty(TOOLTIP.ARROW_BORDER_COLOR, "transparent transparent white transparent");
  }

  let arrowLeft = elRect.left + elRect.width / 2 - arrow.offsetWidth / 2;
  const arrowTop = elRect.bottom;

  if (modifiers.bottomStart || value?.placement === TOOLTIP_PLACEMENTS.BOTTOM_START) {
    arrowLeft = elRect.left + arrow.offsetWidth;
  }

  if (modifiers.bottomEnd || value?.placement === TOOLTIP_PLACEMENTS.BOTTOM_END) {
    arrowLeft = elRect.right - arrow.offsetWidth * 2;
  }

  setPosition(arrow, arrowLeft, arrowTop);
}

function setArrowLeftPosition(
  el: HTMLElement,
  arrow: HTMLElement,
  modifiers: DirectiveBinding["modifiers"],
  value: DirectiveBinding["value"],
) {
  const elRect = getElRect(el);

  arrow.style.setProperty(TOOLTIP.ARROW_BORDER_COLOR, "transparent transparent transparent #1e293b");

  if (modifiers.bgLight || value?.bgLight) {
    arrow.style.setProperty(TOOLTIP.ARROW_BORDER_COLOR, "transparent transparent transparent white");
  }

  const arrowLeft = elRect.left - arrow.offsetWidth;

  let arrowTop = elRect.top + elRect.height / 2 - arrow.offsetHeight / 2;

  if (modifiers.leftStart || value?.placement === TOOLTIP_PLACEMENTS.LEFT_START) {
    arrowTop = elRect.top + arrow.offsetHeight;
  }

  if (modifiers.leftEnd || value?.placement === TOOLTIP_PLACEMENTS.LEFT_END) {
    arrowTop = elRect.bottom - arrow.offsetHeight * 2;
  }

  setPosition(arrow, arrowLeft, arrowTop);
}

function setArrowRightPosition(
  el: HTMLElement,
  arrow: HTMLElement,
  modifiers: DirectiveBinding["modifiers"],
  value: DirectiveBinding["value"],
) {
  const elRect = getElRect(el);

  arrow.style.setProperty(TOOLTIP.ARROW_BORDER_COLOR, "transparent #1e293b transparent transparent");

  if (modifiers.bgLight || value?.bgLight) {
    arrow.style.setProperty(TOOLTIP.ARROW_BORDER_COLOR, "transparent white transparent transparent");
  }

  let arrowTop = elRect.top + elRect.height / 2 - arrow.offsetHeight / 2;
  const arrowLeft = elRect.right;

  if (modifiers.rightStart || value?.placement === TOOLTIP_PLACEMENTS.RIGHT_START) {
    arrowTop = elRect.top + arrow.offsetHeight;
  }

  if (modifiers.rightEnd || value?.placement === TOOLTIP_PLACEMENTS.RIGHT_END) {
    arrowTop = elRect.bottom - arrow.offsetHeight * 2;
  }

  setPosition(arrow, arrowLeft, arrowTop);
}

function alignTop(
  el: HTMLElement,
  container: HTMLDivElement,
  modifiers: DirectiveBinding["modifiers"],
  value: DirectiveBinding["value"],
) {
  const elRect = getElRect(el);

  const left = elRect.left + (elRect.width - container.offsetWidth) / 2;
  const top = elRect.top - container.offsetHeight - 12;

  setPosition(container, left, top);

  if (arrow) {
    setArrowTopPosition(el, arrow, modifiers, value);
  }
}

function alignTopStart(
  el: HTMLElement,
  container: HTMLDivElement,
  modifiers: DirectiveBinding["modifiers"],
  value: DirectiveBinding["value"],
) {
  const elRect = getElRect(el);

  const left = elRect.left;
  const top = elRect.top - container.offsetHeight - 12;

  setPosition(container, left, top);

  if (arrow) {
    setArrowTopPosition(el, arrow, modifiers, value);
  }
}

function alignTopEnd(
  el: HTMLElement,
  container: HTMLDivElement,
  modifiers: DirectiveBinding["modifiers"],
  value: DirectiveBinding["value"],
) {
  const elRect = getElRect(el);

  const left = elRect.right - container.offsetWidth;
  const top = elRect.top - container.offsetHeight - 12;

  setPosition(container, left, top);

  if (arrow) {
    setArrowTopPosition(el, arrow, modifiers, value);
  }
}

function alignBottom(
  el: HTMLElement,
  container: HTMLDivElement,
  modifiers: DirectiveBinding["modifiers"],
  value: DirectiveBinding["value"],
) {
  const elRect = getElRect(el);

  const left = elRect.left + (elRect.width - container.offsetWidth) / 2;
  const top = elRect.bottom + 12;

  setPosition(container, left, top);

  if (arrow) {
    setArrowBottomPosition(el, arrow, modifiers, value);
  }
}

function alignBottomStart(
  el: HTMLElement,
  container: HTMLDivElement,
  modifiers: DirectiveBinding["modifiers"],
  value: DirectiveBinding["value"],
) {
  const elRect = getElRect(el);

  const left = elRect.left;
  const top = elRect.bottom + 12;

  setPosition(container, left, top);

  if (arrow) {
    setArrowBottomPosition(el, arrow, modifiers, value);
  }
}

function alignBottomEnd(
  el: HTMLElement,
  container: HTMLDivElement,
  modifiers: DirectiveBinding["modifiers"],
  value: DirectiveBinding["value"],
) {
  const elRect = getElRect(el);

  const left = elRect.right - container.offsetWidth;
  const top = elRect.bottom + 12;

  setPosition(container, left, top);

  if (arrow) {
    setArrowBottomPosition(el, arrow, modifiers, value);
  }
}

function alignLeft(
  el: HTMLElement,
  container: HTMLDivElement,
  modifiers: DirectiveBinding["modifiers"],
  value: DirectiveBinding["value"],
) {
  const elRect = getElRect(el);

  const left = elRect.left - container.offsetWidth - 12;
  const top = elRect.top + (elRect.height - container.offsetHeight) / 2;

  setPosition(container, left, top);

  if (arrow) {
    setArrowLeftPosition(el, arrow, modifiers, value);
  }
}

function alignLeftStart(
  el: HTMLElement,
  container: HTMLDivElement,
  modifiers: DirectiveBinding["modifiers"],
  value: DirectiveBinding["value"],
) {
  const elRect = getElRect(el);

  const left = elRect.left - container.offsetWidth - 12;
  const top = elRect.top;

  setPosition(container, left, top);

  if (arrow) {
    setArrowLeftPosition(el, arrow, modifiers, value);
  }
}

function alignLeftEnd(
  el: HTMLElement,
  container: HTMLDivElement,
  modifiers: DirectiveBinding["modifiers"],
  value: DirectiveBinding["value"],
) {
  const elRect = getElRect(el);

  const left = elRect.left - container.offsetWidth - 12;
  const top = elRect.bottom - container.offsetHeight;

  setPosition(container, left, top);

  if (arrow) {
    setArrowLeftPosition(el, arrow, modifiers, value);
  }
}

function alignRight(
  el: HTMLElement,
  container: HTMLDivElement,
  modifiers: DirectiveBinding["modifiers"],
  value: DirectiveBinding["value"],
) {
  const elRect = getElRect(el);

  const left = elRect.right + 12;
  const top = elRect.top + (elRect.height - container.offsetHeight) / 2;

  setPosition(container, left, top);

  if (arrow) {
    setArrowRightPosition(el, arrow, modifiers, value);
  }
}

function alignRightStart(
  el: HTMLElement,
  container: HTMLDivElement,
  modifiers: DirectiveBinding["modifiers"],
  value: DirectiveBinding["value"],
) {
  const elRect = getElRect(el);

  const left = elRect.right + 12;
  const top = elRect.top;

  setPosition(container, left, top);

  if (arrow) {
    setArrowRightPosition(el, arrow, modifiers, value);
  }
}

function alignRightEnd(
  el: HTMLElement,
  container: HTMLDivElement,
  modifiers: DirectiveBinding["modifiers"],
  value: DirectiveBinding["value"],
) {
  const elRect = getElRect(el);

  const left = elRect.right + 12;
  const top = elRect.bottom - container.offsetHeight;

  setPosition(container, left, top);

  if (arrow) {
    setArrowRightPosition(el, arrow, modifiers, value);
  }
}

function isOutOfBounds(container: HTMLDivElement, cb: (isOut: boolean) => void) {
  const containerRect = container.getBoundingClientRect();

  const width = container.offsetWidth;
  const height = container.offsetHeight;
  const viewportWidth = window.innerWidth || document.documentElement.clientWidth;
  const viewportHeight = window.innerHeight || document.documentElement.clientHeight;

  cb(
    containerRect.left + width > viewportWidth
    || containerRect.left < 0
    || containerRect.top < 0
    || containerRect.top + height > viewportHeight,
  );
}
