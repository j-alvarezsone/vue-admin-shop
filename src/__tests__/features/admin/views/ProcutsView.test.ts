import type { Mock } from "vitest";
import { fakeProducts } from "@/__tests__/fake/products.fake";
import { tooltip } from "@/directives/tooltip";
import ProductsView from "@/features/admin/views/ProductsView.vue";
import { useQuery, useQueryClient } from "@tanstack/vue-query";
import { shallowMount } from "@vue/test-utils";
import { createPinia } from "pinia";
import { createRouter, createWebHistory } from "vue-router";

const router = createRouter({
  history: createWebHistory(),
  routes: [
    {
      path: "/",
      component: ProductsView,
    },
  ],
});

vi.mock("@tanstack/vue-query", () => {
  return {
    useQueryClient: vi.fn().mockReturnValue({
      prefetchQuery: vi.fn(),
    }),
    useQuery: vi.fn(),
    useMutation: vi.fn().mockReturnValue({ mutate: vi.fn() }),
  };
});

describe("<ProductsView />", () => {
  (useQuery as Mock).mockReturnValue({
    data: fakeProducts,
  });

  window.scrollTo = vi.fn();

  const wrapper = shallowMount(ProductsView, {
    global: {
      plugins: [router, createPinia()],
      directives: {
        tooltip,
      },
    },
  });

  it("should render with default values", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("should prefetch query on mounted", async () => {
    await router.replace("/?page=2");

    expect(useQueryClient().prefetchQuery).toHaveBeenCalledWith({
      queryKey: ["products", { page: 3 }],
      queryFn: expect.any(Function),
      staleTime: 1000 * 30,
    });

    expect(window.scrollTo).toHaveBeenCalledWith({
      behavior: "smooth",
      top: 0,
    });
  });
});
