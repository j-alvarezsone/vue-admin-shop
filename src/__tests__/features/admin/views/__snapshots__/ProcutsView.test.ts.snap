// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<ProductsView /> > should render with default values 1`] = `
"<div class="bg-white px-5 py-2 rounded">
  <div class="flex items-center gap-6">
    <h1 class="text-3xl"> Products </h1>
    <input-text-stub name="search" placeholder="Search by title" required="true" type="text" tooltipplacement="top" icon="search" iconpos="left" size="md" iconsize="size-4" standalone="false" disabled="false" copy="false" noborder="false" loading="false" clearable="true" onepassword="false" modelvalue="" edit="false" class="w-full"></input-text-stub>
    <input-select-stub options="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]" name="size" placeholder="Select size" disabled="false" searchable="false" size="md" clearable="true" searchclearable="true" standalone="false" required="true" loading="false" itemheight="44" tooltipplacement="top" copy="false" disablefilter="false" selectfirst="false" modelvalue="all" search="" class="w-60"></input-select-stub>
  </div>
  <div class="py-8 w-full">
    <div class="shadow overflow-hidden rounded border-b border-gray-200">
      <table class="min-w-full bg-white">
        <thead class="bg-gray-800 text-white">
          <tr>
            <th class="w-10 text-left py-3 px-4 uppercase font-semibold text-sm"> Images </th>
            <th class="flex-1 text-left py-3 px-4 uppercase font-semibold text-sm"> Title </th>
            <th class="w-28 py-3 px-4 uppercase font-semibold text-sm"> Price </th>
            <th class="w-60 text-left py-3 px-4 uppercase font-semibold text-sm"> Sizes </th>
          </tr>
        </thead>
        <tbody class="text-gray-700"></tbody>
      </table>
      <!--v-if-->
    </div>
  </div>
  <pagination-stub page="1" hasmoredata="false" class="bg-white"></pagination-stub>
</div>"
`;
